src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -29309,32 +29325,361 @@ class C<T>
+                // (12,9): error CS8762: Parameter 'z' may not have a null value when exiting with 'false'.
+                //         return y != null || z != null;
+                Diagnostic(ErrorCode.WRN_ParameterConditionallyDisallowsNull, "return y != null || z != null;").WithArguments("z", "false").WithLocation(12, 9)
+                );
+        }
+
+        [Theory]
+        [InlineData("TStruct?")]
+        [InlineData("TClass?")]
+        [WorkItem(39922, "https://github.com/dotnet/roslyn/issues/39922")]
+        public void EnforcedInMethodBody_NotNullWhen(string type)
+        {
+            var source = @"
+#nullable enable
+
+using System.Diagnostics.CodeAnalysis;
+
+class C<TStruct, TClass>
+    where TStruct : struct
+    where TClass : class
+{
+    static bool TryGetValue([NotNullWhen(true)] out TYPE y)
+    {
+        y = null;
+        if (y == null)
+        {
+            return true; // 1
+        }
+        return false;
+    }
+
+    static bool TryGetValue2([NotNullWhen(true)] out TYPE y)
+    {
+        y = null;
+        return y != null;
+    }
+
+    static bool TryGetValue3([NotNullWhen(false)] out TYPE y)
+    {
+        y = null;
+        return y == null;
+    }
+
+    static bool TryGetValue4([NotNull] TYPE x, [NotNullWhen(false)] out TYPE y)
+    {
+        y = null;
+        if (y != null)
+        {
+            return true;
+        }
+        return false; // 2
+    }
+}
+";
+            var comp = CreateNullableCompilation(new[] { NotNullAttributeDefinition, NotNullWhenAttributeDefinition, source.Replace("TYPE", type) });
+            comp.VerifyDiagnostics(
+                // (15,13): error CS8762: Parameter 'y' may not have a null value when exiting with 'true'.
+                //             return true; // 1
+                Diagnostic(ErrorCode.WRN_ParameterConditionallyDisallowsNull, "return true;").WithArguments("y", "true").WithLocation(15, 13),
+                // (39,9): error CS8762: Parameter 'y' may not have a null value when exiting with 'false'.
+                //         return false; // 2
+                Diagnostic(ErrorCode.WRN_ParameterConditionallyDisallowsNull, "return false;").WithArguments("y", "false").WithLocation(39, 9)
+                );
+        }
+
+        [Theory]
+        [InlineData("TStruct?")]
+        [InlineData("TClass?")]
+        [WorkItem(39922, "https://github.com/dotnet/roslyn/issues/39922")]
+        public void EnforcedInMethodBody_NotNullWhen_Composition(string type)
+        {
+            var source = @"
+#nullable enable
+
+using System.Diagnostics.CodeAnalysis;
+
+class C<TStruct, TClass>
+    where TStruct : struct
+    where TClass : class
+{
+    static bool TryGetValue([NotNullWhen(true)] out TYPE y)
+    {
+        return TryGetValue(out y);
+    }
+
+    static bool TryGetValue2([NotNullWhen(true)] out TYPE y)
+    {
+        return TryGetValue3(out y); // 1
```

Comment:
Do we get an analogous warning behavior when a method equivalent to `TryGetValue3` calls `TryGetValue2`?